### DISCUSSION TITLE
Make use of recent API extensions of Apache PLC4X

### DISCUSSION THREAD

**Discussion by chrisdutz** [apache/camel-karavan#1197] [Ideas]
Hi all,

I just wanted to bring to your attention, that in the Apache PLC4X project we just recently did quite some improvements on our API, that now allow tools to introspect a lot more:
- Which protocols are available?
- Which transport options doe a given protocol have?
- Which default transport does a given protocol use?
- Which options does a protocol have?
- Which options are required?
- Which defaults do options have?
- Which options does a transport used with a given protocol support? 
- ...

We did that to make it possible to create something like an editor or wizzard that should allow people to configure PLC4X connections a lot easier. 

Right now I think there's only a text field, accepting a connection string and there is no validation of that or any support on finding if an option was mis-typed or the value has the wrong format. 

Unfortunately I am not too deep into Karavan, so me building a ui to utilize these improvements is probably not going to happen any time soon, but I thought I should bring this up here and I'm happy to be of assistance if anyone want's to implement this.

Chris

**Comment by mgubaidullin** [comment]
Hello @chrisdutz .Appreciate your interest.

Karavan Component configuration properties panel generated from Component metadata catalog, ex:

AciveMQ Component metadata in Apache Camel: https://github.com/apache/camel/blob/main/catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/components/activemq.json generates the following UI 

![active-mq](https://github.com/apache/camel-karavan/assets/1379213/04376fef-219a-423f-9324-0db19c5c7175)






As you could see it supports TextField, Numeric Fields, Boolean Switch, Dropdown, etc. 
 

So to make PLC4X configurable, all options should be added to PLC4X Component.

**Reply by chrisdutz** [reply]
The problem I see here is that I assume the metadata to be static ... in case of PLC4X it's dynamic. So depending on which protocol you chose, you have different choices of transports to use and also have different options and differences in which options are required and which default values some have ... so is there some way to work with dynamic metadata?

**Reply by oscerd** [reply]
One of the possible way to do this is implementing multiple components, one of each driver here: https://github.com/apache/plc4x/tree/develop/plc4j/drivers/

We are doing something like that in camel-kubernetes

**Reply by chrisdutz** [reply]
Yeah ... I would really love to not have to do that ;-) It's quite a number of drivers that we currently support :-/

Perhaps one day, I'll feel the urge to do that, but I guess till then I'll leave it for someone else to take up. 

Unfortunately way too much open tasks in Apache PLC4X to deal with and way to little people working on them.

**Reply by mgubaidullin** [reply]
Another option is to create a Driver Specific Kamelets that reuses the same Component but provides specific configuration properties. Ex. we have a generic SQL Component and specific Kamelets for MySQL, Postgres, etc. 
https://github.com/apache/camel-kamelets/blob/main/kamelets/postgresql-sink.kamelet.yaml

The advantages with Kamelets is that they are simple YAML files, so they might be generated from PLC4X Driver configurations

**Reply by chrisdutz** [reply]
Hmmm ... that might be an option, as I used the API extensions to auto generate the asciidoc documentation of our drivers based on what the API provides ... that might be something worth investigating: https://github.com/apache/plc4x/tree/develop/plc4j/drivers/all/src/site/generated

**Comment by chrisdutz** [comment]
I guess if someone here could help me with how such a config file would need to look for one driver, I could probably automate that.

**Reply by mgubaidullin** [reply]
I have never worked with PLC4X. Can you make a working example with current PLC4X Component in Camel (using Java or XML).
Based on this example I will try to make a Kamelet.

**Reply by mgubaidullin** [reply]
Is this one correct? https://github.com/apache/camel/blob/4279ded2b624f3b2ab9ef16c1a41875247452365/components/camel-plc4x/src/test/java/org/apache/camel/component/plc4x/ManualTest.java#L53

**Reply by chrisdutz** [reply]
It should

**Comment by davsclaus** [comment]
In Camel 4.5 we generate metadata for pojo beans, so you can also explore to have speicific driver beans that can be configured using their own options, and then from the plc4j component choose a drive (or let camel auto-detect if there is only 1 driver bean configured).

The karavan can be enhanced to make configuring pojo beans (beans) in its UI able to use that metadata to show the same form UI that it would do for components etc. Then you have the same power.

There are the out of the box beans we have in camel 4.5
https://github.com/apache/camel/tree/main/catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/beans
